packages/@sanity/form-builder/src/inputs/Slug/utils/useAsync.tsx

```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+
+export type AsyncCompleteState<T> = {
+  status: 'complete'
+  result: T
+}
+export type AsyncPendingState = {
+  status: 'pending'
+}
+export type AsyncErrorState = {
+  status: 'error'
+  error: Error
+}
+
+export type AsyncState<T> = AsyncPendingState | AsyncCompleteState<T> | AsyncErrorState
+
+export function useAsyncCallback<T, U>(
```

Comment:
We should add some TSDoc to describe what this is and what it does, and whether there are any known gotchas.

Reply:
Added some tsdocs to it (and realized it was named wrong). Fixed & pushed. Let's consider if we should lift it over to a generic tool library at some point.